### PR TITLE
LUCENE-8332 move CompletionTokenStream to ConcatenateGraphFilter

### PR DESCRIPTION
**Pull request by dsmiley**
See JIRA issue for overview comments.

### CODE REVIEW

File: lucene/analysis/common/src/java/org/apache/lucene/analysis/miscellaneous/ConcatenateGraphFilter.java [hunk 1]
```diff
@@ -31,80 +33,106 @@
...
- * a completion's payload (see {@link CompletionTokenStream#setPayload(org.apache.lucene.util.BytesRef)})
+ * Concatenates/Joins every incoming token with a separator into one output token for every path through the
+ * token stream (which is a graph).  In simple cases this yields one token, but in the presence of any tokens with
+ * a zero positionIncrmeent (e.g. synonyms) it will be more.  This filter uses the token bytes, position increment,
+ * and position length of the incoming stream.  Other attributes are not used or manipulated.
  *
  * @lucene.experimental
  */
-public final class CompletionTokenStream extends TokenStream {
+public final class ConcatenateGraphFilter extends TokenFilter {
```
Comment (jimczi) [inline lucene/analysis/common/src/java/org/apache/lucene/analysis/miscellaneous/ConcatenateGraphFilter.java]: I still don't understand why you need to use a TokenFilter instead of a TokenStream. You need to override all the functions anyway so I don't see the benefit.

File: lucene/analysis/common/src/java/org/apache/lucene/analysis/miscellaneous/ConcatenateGraphFilter.java [hunk 2]
```diff
@@ -31,80 +33,106 @@
...
+   * {@link org.apache.lucene.analysis.tokenattributes.TermToBytesRefAttribute} attribute
+   * The token stream uses a {@link org.apache.lucene.analysis.tokenattributes.PayloadAttribute} to store
+   * a completion's payload (see {@link ConcatenateGraphFilter#setPayload(org.apache.lucene.util.BytesRef)})
+   */
+
+  /**
+   * Represents the separation between tokens, if
+   * <code>preserveSep</code> is <code>true</code>.
+   */
+  public final static char SEP_CHAR = '\u001F';
```
Comment (jimczi) [inline lucene/analysis/common/src/java/org/apache/lucene/analysis/miscellaneous/ConcatenateGraphFilter.java]: I think it's better to keep the SEP_LABEL as an int. It needs to be equal to TokenStreamToAutomaton.POS_SEP anyway so you can maybe used this value directly ? PAYLOAD_SEP has the same requirement so we should also set it to POS_SEP.

File: lucene/analysis/common/src/java/org/apache/lucene/analysis/miscellaneous/ConcatenateGraphFilter.java [hunk 3]
```diff
@@ -119,27 +147,19 @@ public boolean incrementToken() throws IOException {
...
+    restoreState(endState);
   }
 
+  //nocommit move method to before incrementToken
   @Override
   public void reset() throws IOException {
-    super.reset();
-    if (hasAttribute(CharTermAttribute.class)) {
-      // we only create this if we really need it to safe the UTF-8 to UTF-16 conversion
-      charTermAttribute = getAttribute(CharTermAttribute.class);
```
Comment (jimczi) [inline lucene/analysis/common/src/java/org/apache/lucene/analysis/miscellaneous/ConcatenateGraphFilter.java]: The automaton should be created lazily on the first call to incrementToken like described in the comment above ;).

File: lucene/analysis/common/src/java/org/apache/lucene/analysis/miscellaneous/ConcatenateGraphFilterFactory.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
+/*
+ This software was produced for the U. S. Government
+ under Contract No. W15P7T-11-C-F600, and is
+ subject to the Rights in Noncommercial Computer Software
+ and Noncommercial Computer Software Documentation
+ Clause 252.227-7014 (JUN 1995)
+
```
Comment (jimczi) [inline lucene/analysis/common/src/java/org/apache/lucene/analysis/miscellaneous/ConcatenateGraphFilterFactory.java]: Is this really needed ? I understand that you have some constraint for the development of the main feature but this doesn't seem related so maybe use the usual license ?

File: lucene/suggest/src/java/org/apache/lucene/search/suggest/document/NRTSuggesterBuilder.java [hunk 5]
```diff
@@ -43,6 +44,9 @@
    * in the output
    */
   public static final int PAYLOAD_SEP = '\u001F';
```
Comment (jimczi) [inline lucene/suggest/src/java/org/apache/lucene/search/suggest/document/NRTSuggesterBuilder.java]: You can set it to TokenStreamToAutomaton.POS_SEP directly and add a test to ensure that PAYLOAD_SEP and SEP_LABEL are equals to `0x001F` for BWC.

File: lucene/suggest/src/java/org/apache/lucene/search/suggest/document/NRTSuggesterBuilder.java [hunk 6]
```diff
@@ -43,6 +44,9 @@
...
+  static {
+    assert PAYLOAD_SEP == ConcatenateGraphFilter.SEP_CHAR;
```
Comment (jimczi) [inline lucene/suggest/src/java/org/apache/lucene/search/suggest/document/NRTSuggesterBuilder.java]: This can ensured in a dedicated test ?

File: lucene/analysis/common/src/java/org/apache/lucene/analysis/miscellaneous/ConcatenateGraphFilterFactory.java (hunk 4 above)
Comment (dsmiley) [reply lucene/analysis/common/src/java/org/apache/lucene/analysis/miscellaneous/ConcatenateGraphFilterFactory.java]: Good catch; no it's not needed.

File: lucene/suggest/src/java/org/apache/lucene/search/suggest/document/NRTSuggesterBuilder.java (hunk 6 above)
Comment (dsmiley) [reply lucene/suggest/src/java/org/apache/lucene/search/suggest/document/NRTSuggesterBuilder.java]: I think it's useful to read the code and both see what's it's value is and that it's equal to something else without having to navigate or use IDE functions.  But this is no big deal to me; I can put back.  Presumably if these values were not the same then some a test would fail but I'm not sure.

File: lucene/analysis/common/src/java/org/apache/lucene/analysis/miscellaneous/ConcatenateGraphFilter.java (hunk 2 above)
Comment (dsmiley) [reply lucene/analysis/common/src/java/org/apache/lucene/analysis/miscellaneous/ConcatenateGraphFilter.java]: I think this is the difference between an implementation detail and something more meaningful/intuitive to the user.  I'm trying to cater to the user here.  Maybe we could have both?

File: lucene/analysis/common/src/java/org/apache/lucene/analysis/miscellaneous/ConcatenateGraphFilter.java (hunk 3 above)
Comment (dsmiley) [reply lucene/analysis/common/src/java/org/apache/lucene/analysis/miscellaneous/ConcatenateGraphFilter.java]: I understand what you are asking me to do (to put back) but don't understand _why_.  What comment above do you refer to?  Where your feedback is?  That comment doesn't reference that; or at least it's very non-obvious to me if there's a relationship (and we should add comments explaining).

File: lucene/analysis/common/src/java/org/apache/lucene/analysis/miscellaneous/ConcatenateGraphFilter.java (hunk 1 above)
Comment (dsmiley) [reply lucene/analysis/common/src/java/org/apache/lucene/analysis/miscellaneous/ConcatenateGraphFilter.java]: Perhaps it _never_ matters; yet TokenFilter exists nonetheless.  TokenFilter is more consistent with public/visible analysis components, and it makes it easier to pass-through the calls to the delegate/input.  

Let me put it another way... shouldn't FingerprintFilter & ConcatenateGraphFilter do the same, whatever that is?  FingerprintFilter is already public/known so I'm trying to introduce this in a way that is consistent.  That CompletionTokenStream, kinda hidden, is currently not a TokenFilter isn't relevant.

File: lucene/analysis/common/src/java/org/apache/lucene/analysis/miscellaneous/ConcatenateGraphFilter.java (hunk 3 above)
Comment (jimczi) [reply lucene/analysis/common/src/java/org/apache/lucene/analysis/miscellaneous/ConcatenateGraphFilter.java]: I am referring to this comment: 
````
//nocommit move method to before incrementToken
````
but maybe I misunderstood what you meant by move. Let me ask you this differently, why do you need to build the automaton on reset ? Unless there is a good reason to do it we shouldn't change the behavior here which is to build the automaton on the first call to incrementToken.

File: lucene/analysis/common/src/java/org/apache/lucene/analysis/miscellaneous/ConcatenateGraphFilter.java (hunk 2 above)
Comment (jimczi) [reply lucene/analysis/common/src/java/org/apache/lucene/analysis/miscellaneous/ConcatenateGraphFilter.java]: This value should only be used internally to encode a separator in the automaton or in the binary terms that we index. Which user are you referring to ?

File: lucene/analysis/common/src/java/org/apache/lucene/analysis/miscellaneous/ConcatenateGraphFilter.java (hunk 3 above)
Comment (dsmiley) [reply lucene/analysis/common/src/java/org/apache/lucene/analysis/miscellaneous/ConcatenateGraphFilter.java]: Oh *that* comment.  By that I mean move the _method_ to follow a natural sequence.  I like classes to have methods that are organized sensibly.  When one uses a TokenStream, reset() should come before incrementToken() !  -- then end(), then close(), and any utility methods should come after callers.

I moved the initialization to reset simply because it felt natural this way rather than incrementToken where it's lazy.  That's all.  That said... I hinted at a TestRandomChains failure about lifecycle needing to throw an IllegalStateException or something like that (pertaining to missing close()), and I think making this back to lazy will "fix" that.

File: lucene/analysis/common/src/java/org/apache/lucene/analysis/miscellaneous/ConcatenateGraphFilter.java (hunk 2 above)
Comment (dsmiley) [reply lucene/analysis/common/src/java/org/apache/lucene/analysis/miscellaneous/ConcatenateGraphFilter.java]: One user is the SolrTextTagger which needs to know the character separator between concatenated terms.  https://github.com/OpenSextant/SolrTextTagger/blob/master/src/main/java/org/opensextant/solrtexttagger/TermPrefixCursor.java#L45   (being added to Solr in SOLR-12376).  Hmmm, though it is a byte there, which is to your point.  Maybe I should change to a byte and add a little comment about this?

File: lucene/analysis/common/src/java/org/apache/lucene/analysis/miscellaneous/ConcatenateGraphFilter.java (hunk 1 above)
Comment (jimczi) [reply lucene/analysis/common/src/java/org/apache/lucene/analysis/miscellaneous/ConcatenateGraphFilter.java]: Converting this token stream into a token filter is trappy. It requires to rewrite the logic completely. For instance you removed the `close` implementation, this means that it is called twice on the input stream since it is already closed here https://github.com/apache/lucene-solr/pull/384/files#diff-04d4c6352889dbffd0eb4d1e6ecd6097R197. You also added a call to super() which was avoided clearly in the original impl:
````
// Don't call the super(input) ctor - this is a true delegate and has a new attribute source since we consume
    // the input stream entirely in the first call to incrementToken
````
My idea was to move this token stream if only minor changes are required. If it needs a complete rewrite then we should probably reconsider.

Comment (jimczi) [reply lucene/analysis/common/src/java/org/apache/lucene/analysis/miscellaneous/ConcatenateGraphFilter.java]: +1

**Comment by dsmiley** [comment]
There's a test failure in TestFactories.test which calls BaseTokenStreamTestCase.checkResetException.

`ant test -Dtestcase=TestFactories -Dtests.seed=543C55AE2375085C`
```
   [junit4]    > Throwable #1: java.lang.AssertionError: didn't get expected exception when close() not called
   [junit4]    >        at __randomizedtesting.SeedInfo.seed([543C55AE2375085C:DC686A748D8965A4]:0)
   [junit4]    >        at org.apache.lucene.analysis.BaseTokenStreamTestCase.checkResetException(BaseTokenStreamTestCase.java:453)
   [junit4]    >        at org.apache.lucene.analysis.BaseTokenStreamTestCase.checkRandomData(BaseTokenStreamTestCase.java:546)
   [junit4]    >        at org.apache.lucene.analysis.core.TestFactories.doTestTokenFilter(TestFactories.java:118)
   [junit4]    >        at org.apache.lucene.analysis.core.TestFactories.test(TestFactories.java:68)
   [junit4]    >        at java.lang.Thread.run(Thread.java:748)
```
The issue is that if some faulty code forgets to call close(), an exception would be ideal to be thrown but for this filter it won't be because the inputTokenStream is closed properly internally correctly by toAutomaton.  Checks like this are kinda... I dunno, I don't like it.  IMO it's asking too much of a TokenFilter to insist that it's cranky when it's mistreated; maybe some components (like this one) are more resilient?  Maybe BaseTokenStreamTestCase (or somewhere else) should keep track of components that are resilient in this way so that it doesn't exercise these sort of tests?  If we really want to insist that every component be cranky when mistreated, then perhaps we could have toAutomaton _not_ close the inputTokenStream so that we can delegate that from CGF's close?

**Comment by jimczi** [comment]
> If we really want to insist that every component be cranky when mistreated, then perhaps we could have toAutomaton not close the inputTokenStream so that we can delegate that from CGF's close?

I think it's the right thing to do and the change is minimal. Good catch !
